Service loopback

tgraf opened the pull request.

aanm reviewed (commented):
>Note: IPv4 only implementation

Does this happens with IPv6?

> FIXME: Needs a dedicated IPv4 address to ensure packets from the local host can never cause clashing tuples.

Where is this IP needed? Can you provide more info regarding this?

tgraf committed:
bpf: Provide hints for BPF context offsets

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: All loopback back to origin via service IP (IPv4)

endpointA => service IP => endpointA caused packets to be dropped due
to Linux martian source detection. Perform special SNAT for this
particular loopback case to avoid a martian source address.

Reserves a special IPv4 address out of the alloc range for this
purpose to gurantee unique tuples.

Signed-off-by: Thomas Graf <thomas@cilium.io>

aanm commented on daemon/daemon/daemon.go:
```diff
@@ -422,6 +424,13 @@ func NewDaemon(c *Config) (*Daemon, error) {
 		return nil, err
 	}
 
+	// Allocate IPv4 service loopback IP
```
Put this condition `if d.conf.IPv4Enabled`
